core/src/commonMain/kotlin/Characteristic.kt

```diff
@@ -0,0 +1,27 @@
+package com.juul.kable
+
+import com.benasher44.uuid.Uuid
+import com.benasher44.uuid.uuidFrom
+
+public enum class WriteType {
+    WithResponse,
+    WithoutResponse,
+}
+
+public expect class Characteristic {
+    public val uuid: Uuid
+    public val descriptors: List<Descriptor>
+}
+
+public operator fun List<Characteristic>?.get(
```

Comment:
> in my experience it's usually preferred to push the nullable up out of the API to a higher level.

Are you thinking that this should instead be non-`null` and throw the exception when we try to request a non-existent (or non-discovered) characteristic? Then at a higher level we can catch that exception and return `null` instead (if desired)?

Reply:
I wasn't very clear describing what I mean by "push the nullable up", sorry about that. And apologies in advance if this is now too much bike-shedding. I blame COVID.

Just looking at the shape of this function in isolation i.e.:
```kotlin
fun List<Characteristic>?.get(uuid: String): Characteristic
```

For this to be a useful function signature, it means somebody has to have a variable of type `List<Characteristic>?` on which they may invoke the `get` method, right? e.g.:
```kotlin
var myList: List<Characteristic>?
...
var foo = myList.get(...)
```

In my mind it makes more sense to push the nullability responsibility (ha rhymes!) up to the owner where, if they do insist on using a nullable type, they can use `map` (or whatever the equivalent safe dereference operator is called) prior to invoking the accessor e.g.
```kotlin
fun List<Characteristic>.get(uuid: String): Characteristic
...
var myList: List<Characteristic>?
...
var foo = myList?.get(...)
```

On the face of it, it looks like the shape of this function is forcing the caller to embrace nullability in their variable declarations. Perhaps that is actually desired for the use case, and if so that is cool.

Reply:
I'm wondering if we'd be better off with a custom container class here, instead of adding extension functions to `List<Characteristic>` in the first place. Seems like the data structure shouldn't be intrinsic to the logic (I could easily see a world where we want a `Map<Uuid, Characteristic>` as the backing structure instead).

Reply:
A custom container class could do the trick, though I wonder if it might add unnecessary complexity?

Let me explain my rationale for this (admittedly awkward) API design choice.

`Peripheral.services` property is nullable so that `null` can represent an unknown state (i.e. services have not yet been discovered).

To make the usage of the API more readable, a nullable receiver was chosen so that code like the following could be written:

```kotlin
val characteristic = peripheral
    .services["f000aa80-0451-4000-b000-000000000000"]
    .characteristics["f000aa81-0451-4000-b000-000000000000"]
```

In this example code, `characteristic` will be `null` if services are not yet discovered and throw if a specific UUID is not found; making the code readable and easy to differentiate between the failure cases.

Alternatively I could have `services` property not be nullable and throw when accessed prior to service discovery, but:

> Kotlin encourages writing code in a functional style. It works well as long as business-specific failures are represented with nullable types or sealed class hierarchies

https://github.com/Kotlin/KEEP/blob/master/proposals/stdlib/result.md#functional-bulk-manipulation-of-failures

---

Perhaps I'm trying too hard to make the usage look so clean (at the expense of API understandability)? Or maybe I'm overlooking better API design that still provides the clean usage?

Reply:
Thanks for explaining, I get the reasoning behind it now, it's to allow the syntactic sugar of the index operator `[]` vs the explicit `?.get()` call where the receiver is nullable.

The idea of using a custom container is a more precise/meaningful syntactic sugar. Neither of these get to the underlying issue of the additional state implied by the null.

If the receiver is not nullable none of this is necessary. This is where my anti-null stance comes from in this instance, triggered by the smell this API gives off.

> Alternatively I could have services property not be nullable and throw when accessed prior to service discovery

This gets to the heart of the matter of how to represent the "pre-discovery" state.

Consider if it is nullable, it is null, the user attempts to obtain the characteristic and it returns a null because it is null (regardless if there is or isn't syntactic sugar).

Now consider it is non-nullable, it is an empty list, the user attempts to obtain the characteristic and it returns a null because it is empty.

In order to differentiate the two identical outcomes, the user has to then go "wait a minute, was that thing null? oh yeah it was that means we are in pre-discovery state". In the latter case an error has to be thrown to convey the state information. I am on the fence if this is better or worse - depends whether you come down on the side of providing more or fewer guard rails (forcing an error to convey state, or letting them choose to do a null check to obtain state).

> Kotlin encourages writing code in a functional style

Welp I guess I disagree with the language designers' definition of "function style" in this case. In a functional language with an algebraic type system this concept would be represented explicitly, and you can simulate that functional style directly into Kotlin with a sealed class (albeit syntactically verbose):

```kotlin
sealed class MaybeCharacteristics {
    object PreDiscovery: MaybeCharacteristics() {}

    data class Discovered: MaybeCharacteristics() {
        val characteristics: List<Characteristic>
    }
}
```

I'm not encouraging this in place of `null` because the language is awkward with this style and the KEEP is recommending the use of `null` instead. But "functional style" - you (the KEEP authors, not the people on this PR) keep using that word, I do not think it means what you think it means.

To conclude our Friday bike-shed, I recommend keep the code as if with the optional and the overloaded null state and associated syntax accessors because it works best with the language and is easy enough API to follow provided the meaning of the null state is documented. When I see the type `List<Characteristic>?` in my head I automatically picture the above type information but with `null` in place of `PreDiscovery`. Probably I'm just weird.   😬 

Approved! ✅

Reply:
Personally, I think the cleanest solution here is to separate the `services` collection from the function for getting a single service.

```kotlin
val characteristic = peripheral
    .getService("f000aa80-0451-4000-b000-000000000000")
    ?.getCharacteristic("f000aa81-0451-4000-b000-000000000000")
```

This way we can maintain freedom to use a map or similar, behind the scenes (the `.services` list could be backed simply by `thatMap.values`), while still providing a fairly clean interface.

An alternative, would be services as a wrapper object, such that it always returns non-null, but an additional function is required to listify it.

```kotlin
val characteristic = peripheral
    .services["f000aa80-0451-4000-b000-000000000000"]
    ?.characteristics["f000aa81-0451-4000-b000-000000000000"]

// When needing a list of services
peripheral.services.asList()
```

Reply:
Ultimately decided to leave out the `List` based extension functions.

> When in doubt, leave it out

https://youtu.be/heh4OeB9A-c?t=1474